src/main/java/org/spongepowered/common/event/SpongeCommonEventFactory.java

```diff
@@ -281,6 +284,36 @@ public static CollideEntityEvent callCollideEntityEvent(net.minecraft.world.Worl
         return event;
     }
 
+    /**
+     * This simulates the blocks a piston moves and calls the event for saner debugging.
+     * @return if the event was cancelled
+     */
+    public static boolean handlePistonEvent(IMixinWorldServer world, WorldServer.ServerBlockEventList list, Object obj, BlockPos pos, Block blockIn, int eventId, int eventParam) {
+        boolean extending = (eventId == 0);
+        final IBlockState blockstate = ((net.minecraft.world.World)world).getBlockState(pos);
+        EnumFacing direction = (EnumFacing) blockstate.getValue(BlockDirectional.FACING);
+
+        List<BlockPos> blocks = new ArrayList<BlockPos>();
+        blocks.add(pos);
+
+        BlockPistonStructureHelper movedBlocks = new BlockPistonStructureHelper((WorldServer) (Object) world, pos, direction, extending);
+        movedBlocks.canMove(); // calculates blocks to be moved
+        blocks.addAll(movedBlocks.getBlocksToDestroy());
+        blocks.addAll(movedBlocks.getBlocksToMove());
+
+        List<Location<org.spongepowered.api.world.World>> locations = new ArrayList<Location<org.spongepowered.api.world.World>>();
```

Comment:
And then here:
```java
for (BlockPos pos : movedBlocks.getBlocksToDestroy()) {
  locations.add(new Location<>((World) world, block.getX(), block.getY(), block.getZ());
}
for (BlockPos pos : movedBlocks.getBlocksToMove()) {
  boolean isMatched = false;
  for (Location<World> location : locations) {
    final Vector3i position = location.getBlockPosition();
    if (position.getX() == pos.getX() && position.getY() == pos.getY() && position.getZ() == pos.getZ()) {
      isMatched = true;
      break;
    }
  }
  if (!isMatched) {
    locations.add(new Location<>((World) world, pos.getX(), pos.getY(), pos.getZ());
  }
}
```

Sure, it does two iterations over, but the simple filter check on the blocks to move ensures we don't have duplicate block positions in the first place.